File: packages/sdk-ts/src/instrument/_openai.ts
/** Patch the OpenAI SDK's chat.completions.create + responses.create. */

import { type Unpatch, wrapMethod } from './_common.js';

// biome-ignore lint/suspicious/noExplicitAny: foreign SDK shape
export function instrumentOpenAI(openai: any): Unpatch | null {
  if (!openai) return null;
  const unpatchers: Unpatch[] = [];

  // OpenAI's TS SDK exposes resource classes a few ways depending on
  // version: as `Chat.Completions`, as `resources.chat.completions.
  // Completions`, or as bare exports. Try each.
  const Resources = {
    chatCompletions: openai?.Chat?.Completions ?? openai?.resources?.chat?.completions?.Completions,
    responses: openai?.Responses ?? openai?.resources?.responses?.Responses,
  };

  if (Resources.chatCompletions?.prototype) {
    unpatchers.push(
      wrapMethod(Resources.chatCompletions.prototype, 'create', {
        spanName: 'openai.chat.completions',
        system: 'openai',
        extractModel: (args) => String((args[0] as { model?: string })?.model ?? ''),
        extractMessages: (args) => (args[0] as { messages?: unknown })?.messages,
        extractUsage: (r) => {
          const u = r?.usage ?? {};
          return {
            input_tokens: u.prompt_tokens ?? u.input_tokens,
            output_tokens: u.completion_tokens ?? u.output_tokens,
          };
        },
      }),
    );
  }

  if (Resources.responses?.prototype) {
    unpatchers.push(
      wrapMethod(Resources.responses.prototype, 'create', {
        spanName: 'openai.responses',
        system: 'openai',
        extractModel: (args) => String((args[0] as { model?: string })?.model ?? ''),
        extractMessages: (args) => (args[0] as { input?: unknown })?.input,
        extractUsage: (r) => {
          const u = r?.usage ?? {};
          return { input_tokens: u.input_tokens, output_tokens: u.output_tokens };
        },
      }),
    );
  }

  if (unpatchers.length === 0) return null;
  return () => {
    for (const u of unpatchers) u();
  };
}

File: packages/sdk-python/src/judge/instrument/_openai.py
"""Patch the `openai` SDK's chat-completion + responses methods.

Supports the v1 OpenAI Python SDK (the `openai>=1.0` rewrite). Older
v0 (`openai.ChatCompletion.create`) is not patched here — bump SDKs.
"""

from __future__ import annotations

from typing import Any

from judge.instrument._common import wrap_sync

_unpatchers: list[Any] = []


def _request_model(kwargs: dict[str, Any]) -> str:
    return str(kwargs.get("model", ""))


def _request_messages(kwargs: dict[str, Any]) -> Any:
    # Both chat.completions and responses APIs use `messages` / `input`.
    return kwargs.get("messages") or kwargs.get("input")


def _usage(response: Any) -> dict[str, int]:
    """Pull token counts from an OpenAI response object."""
    u = getattr(response, "usage", None)
    if u is None:
        return {}
    out: dict[str, int] = {}
    in_tokens = getattr(u, "prompt_tokens", None) or getattr(u, "input_tokens", None)
    out_tokens = getattr(u, "completion_tokens", None) or getattr(u, "output_tokens", None)
    if in_tokens is not None:
        out["input_tokens"] = int(in_tokens)
    if out_tokens is not None:
        out["output_tokens"] = int(out_tokens)
    return out


def install() -> bool:
    """Patch OpenAI client classes. Returns True if patched, False if
    `openai` is not importable."""
    try:
        from openai.resources.chat.completions import (  # type: ignore[import-not-found]
            Completions as ChatCompletions,
        )
    except Exception:
        return False

    _unpatchers.append(
        wrap_sync(
            ChatCompletions,
            "create",
            span_name="openai.chat.completions",
            system="openai",
            extract_model=_request_model,
            extract_messages=_request_messages,
            extract_usage=_usage,
        )
    )

    # Newer SDKs also expose `Responses`; patch when present.
    try:
        from openai.resources.responses import (  # type: ignore[import-not-found]
            Responses,
        )

        _unpatchers.append(
            wrap_sync(
                Responses,
                "create",
                span_name="openai.responses",
                system="openai",
                extract_model=_request_model,
                extract_messages=_request_messages,
                extract_usage=_usage,
            )
        )
    except Exception:
        pass

    return True


def uninstall() -> None:
    while _unpatchers:
        _unpatchers.pop()()
